Move default C flags to the `:standard` set

voodoos opened the pull request:
This PR implements RFC #3718 

`ocamlc_cflags` and `ocamlc_cppflags` from `ocamlc_config` are now part of the `:standard` set of flags (it was already the case for the first).
Moreover they are not systematically added any-more to the compiler command line, giving back entire control over c-flags to the users.

These changes may break projects which made the assumption that dune (or ocamlc) added these flags to the command line, which might be considered a bad practice. The change in this PR will activate only when the option `new_foreign_flags_handling` is activated in the dune-project ~dune lang >= 2.8 is used~.

~@jeremiedimino, @rgrinberg : an other compatibility path would be to add an option to the dune-project and only default this behaviour in dune 3. However I believe it might be a good thing to push users who forgot to keep the `:standard` flags to update their projects' configurations.~

rgrinberg commented:
> an other compatibility path would be to add an option to the dune-project and only default this behaviour in dune 3. However I believe it might be a good thing to push users who forgot to keep the :standard flags to update their projects' configurations.

That might be our only option. Otherwise, users can't freely upgrade to `(lang dune 2.8)`

jeremiedimino commented:
I agree with Rudi.

> However I believe it might be a good thing to push users who forgot to keep the :standard flags to update their projects' configurations.

Agreed. Perhaps we can introduce a warning for this?

emillon commented on src/dune_rules/super_context.ml:
```diff
@@ -2,13 +2,16 @@ open! Dune_engine
 open Import
 
-let default_context_flags (ctx : Context.t) =
-  (* TODO DUNE3 To ensure full backward compatibility, ocaml_cflags are still
```
I think there should be a TODO DUNE3 comment somewhere about making the opt-in behavior added in the PR a dune 3 default.

emillon commented on test/blackbox-tests/test-cases/variables/var-cc.t/run.t:
```diff
@@ -2,12 +2,14 @@ According to the doc: CC is the C compiler command line (list made of the
-The flags are made of the :standard (= ocamlc_cflags) set of ocamlc_flags
-merged with (and sometimes replaced by) the flags in the env stanza.
+In practice, in dune < 2.8 it consists in the concatenation of OCaml's
+`c_compiler` and flags The flags are made of the :standard (= ocamlc_cflags) set
```
there seems to be a newline missing here

emillon commented on doc/dune-files.rst:
```diff
@@ -285,6 +285,23 @@ language: The syntax is as a list of the following elements:
+
+.. code:: scheme
+
+    (new_foreign_flags_handling true)
```
I don't think that the name conveys a good description of what it does. An explicit name would be better to inform the user of what it does. And consider the possibility that there'll be a third option in the future - we don't want it to be `even_newer_flags_handling` :) For the same reason, something other than a boolean would be more future proof.

Let me suggest `(foreign_flags_type osl)` (as opposed to `(foreign_flags_type classic)`) maybe?

emillon commented on src/dune_rules/foreign_rules.ml:
```diff
@@ -75,16 +75,21 @@ let include_dir_flags ~expander ~dir (stubs : Foreign.Stubs.t) =
-          ; Fdo.c_flags ctx
-          ]
+        if Dune_project.new_foreign_flags_handling project then
+          Fdo.c_flags ctx
```
Why are FDO flags always passed? is there a test that covers that?

voodoos replied:
They are not always passed, the function may return the empty lsit:
```
let [Fdo.]c_flags (ctx : Context.t) =
  match ctx.fdo_target_exe with
  | None -> []
  | Some _ -> [ "-ffunction-sections" ]
```

emillon commented:
Thanks for this PR.

I think that #3718 makes a good job at describing where flags are sourced from. I think a text description of this would be a nice addition to the manual.

I also second @rgrinberg 's suggestion to have a way for configurator to use the same flags.

rgrinberg commented on src/dune_engine/dune_project.ml:
```diff
@@ -791,6 +800,7 @@ let parse ~dir ~lang ~opam_packages ~file =
      let generate_opam_files =
        Option.value ~default:false generate_opam_files
      in
+     let always_add_cflags = Option.value ~default:true always_add_cflags in
```
I would suggest to pick the default based on the dune version. It should be false for 3.0 and above.

rgrinberg commented on src/dune_rules/super_context.ml:
```diff
@@ -2,13 +2,17 @@ open! Dune_engine
+  in
+  let c =
+    if not (Dune_project.always_add_cflags project) then
+      (* TODO DUNE3 make this the default behavior *)
```
Is this TODO necessary? If this default behavior, this would be reflected by `always_add_cflags` already.

rgrinberg commented on src/dune_engine/dune_project.ml:
```diff
@@ -610,6 +615,7 @@ let infer ~dir packages =
   ; dune_version = lang.version
   ; allow_approx_merlin = true
   ; generate_opam_files = false
+  ; always_add_cflags = true
```
This default should be based on the dune version.

rgrinberg commented:
If we make the change to configurator, how do we version it?

If we make the change to configurator, we need to be mindful of the behavior of:

* `(always_add_cflags false)`, configurator >= 2.8
* `(always_add_cflags false)`, configurator < 2.7

It seems like our hands are tied, since the behavior is always going to be dependent on the version of configurator being used.

The simplest way is to just version people's programs. For example, change Configurator.create to:

```
type cflags =
  | Always_add
  | Ask_dune
val create :
     ?dest_dir:string
  -> ?ocamlc:string
  -> ?log:(string -> unit)
  -> ?cflags:cflags
  -> string (** name, such as library name *)
  -> t
```

That's super safe, but will mean that upgrading is going to be a long process.

jeremiedimino commented:
@rgrinberg we just had a chat about this with @voodoos. The change you propose to confiugurator makes sense to me, however they are not straightforward. IIUC, they require Dune communicating the default C flags as customised by the user in a given directory to configurator, and Dune tracking this dependency so that it re-runs configurator if the user changes the C flags. This is something we currently don't know how to do.

Moreover, were planning to look at the more general question of "dune configure" next.

So let's skip the configurator integration for now, but we'll keep all that in mind when working on the design of "dune configure".

voodoos commented:
I added a warning that triggers if all the following condition are true:
- Dune lang is >= 2.8
- The new `always_add_cflags` is not defined
- The `:standard` set of flags was overriden

The goal is to raise awareness about Dune's forced c-flags and guide the users toward the new default which will be `always_add_cflags false`.

voodoos commented:
I just rebased on master and renamed the option to match the choice done in the twin PR #3802 : `future_c_and_cxx_flags_handling`.

I feel like this PR is ready to be merged, do you agree @rgrinberg, @jeremiedimino ?
If I merge it we will need to merge #3802 before the next release.

rgrinberg commented on doc/dune-files.rst:
```diff
@@ -285,6 +285,23 @@ language: The syntax is as a list of the following elements:
+
+.. code:: scheme
+
+    (future_c_and_cxx_flags_handling true)
```
I have to say that I don't really like this name. I'd prefer a name that describes that hints at what the feature is doing, rather than forecasting that it's going to be a default one day. What about something like `explicit_c_and_cxx_flags_handling` or `c_and_cxx_standard_flags_include_all`?

voodoos replied:
Or maybe `use_standard_c_and_cxx_flags`?

rgrinberg replied:
Sounds like a much better name to me.

voodoos replied:
@jeremiedimino, @emillon, do you agree with this name `use_standard_c_and_cxx_flags` ?

If so I will rename every occurrence.

emillon replied:
I agree that the name is better, yes. I'd rather not use a boolean if possible, but this works too.

rgrinberg replied:
> I agree that the name is better, yes. I'd rather not use a boolean if possible, but this works too.

I also much prefer `on`/`off` or `enabled`/`disabled`. I don't think we have any consistency on this stuff however.

rgrinberg commented on src/dune_rules/foreign_rules.ml:
```diff
@@ -75,19 +75,51 @@ let include_dir_flags ~expander ~dir (stubs : Foreign.Stubs.t) =
+    let has_standard =
+      Ordered_set_lang.Unexpanded.encode flags
+      |> List.map ~f:Dune_lang.to_string
+      |> List.exists ~f:(String.equal ":standard")
```
Could you add an `has_standard` function that works directly on the ast? This is consistent with how we do `is_standard` as well.

rgrinberg commented on src/dune_rules/foreign_rules.ml:
```diff
@@ -75,19 +75,51 @@ let include_dir_flags ~expander ~dir (stubs : Foreign.Stubs.t) =
+    if
+      Dune_project.dune_version project >= (2, 8)
+      && Option.is_none (Dune_project.future_c_and_cxx_flags_handling project)
+      && not has_standard
```
I like this warning, but I'm wondering if we should not emit it for vendored dirs? what do you think?

voodoos replied:
I agree. In fact, shouldn't all warnings be muted for vendored dirs ?

What is the good way to test for vendoring ?

voodoos replied:
I looked into the code and found several duplicate of this process.
I added a function to `File_tree`.

voodoos replied:
https://github.com/ocaml/dune/pull/3875/commits/88be602064d5b8ebeb8390bbdf66d58f190969e6

rgrinberg commented on doc/dune-files.rst:
```diff
@@ -285,6 +285,23 @@ language: The syntax is as a list of the following elements:
+.. _always-add-cflags:
+
+future_c_and_cxx_flags_handling
+-----------------
```
This heading marker should extend to the length of the title. Otherwise sphinx will complain

rgrinberg reviewed (approved):
Looks good apart from some nits. Let's pick a better name as well.

rgrinberg commented:
We handle warnings in quite an adhoc way unfortunately. It's something that I'd like to improve when I get the chance. Feel free to work on it if you're interested

voodoos commented:
This PR is now ready however we are waiting for it's CXX sibling: #3528 
I changed the target milestone to 2.9.

voodoos commented:
@ejgallego same as #3802
